iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,490 @@
+    /// </summary>
+    /// <param name="twinProperties">The TwinProperties object to convert</param>
+    /// <returns></returns>
+    public static Properties FromTwinProperties(TwinProperties twinProperties)
+    {
+        if (twinProperties == null)
+        {
+            throw new ArgumentNullException(nameof(twinProperties));
+        }
+
+        var writablePropertyCollection = new PropertyCollection();
+        foreach (KeyValuePair<string, object> property in twinProperties.Desired)
+        {
+            writablePropertyCollection.AddPropertyToCollection(property.Key, property.Value);
+        }
+
+        var propertyCollection = new PropertyCollection();
+        foreach (KeyValuePair<string, object> property in twinProperties.Reported)
+        {
+            propertyCollection.AddPropertyToCollection(property.Key, property.Value);
+        }
+
+        return new Properties(writablePropertyCollection, propertyCollection);
+    }
+}
+```
+</details>
+
+<details>
+<summary>PropertyCollection</summary>
+
+```csharp
+/// <summary>
+///
+/// </summary>
+public class PropertyCollection : IEnumerable<object>
+{
+    private readonly string _propertyJson;
+    private readonly IDictionary<string, dynamic> _propertiesList = new Dictionary<string, dynamic>();
+
+    internal PropertyCollection()
+    {
+    }
+
+    /// <summary>
+    ///
+    /// </summary>
+    /// <param name="propertyJson"></param>
+    public PropertyCollection(string propertyJson)
+    {
+        _propertyJson = propertyJson;
+    }
+
+    /// <summary>
+    ///
+    /// </summary>
+    /// <param name="propertyName"></param>
+    /// <returns></returns>
+    public dynamic this[string propertyName]
+    {
+        get
+        {
+            return _propertiesList[propertyName];
+        }
+    }
+
+    /// <summary>
+    ///
+    /// </summary>
+    /// <returns></returns>
+    public IEnumerator<object> GetEnumerator()
+    {
+        foreach (object property in _propertiesList)
+        {
+            yield return property;
+        }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+
+    internal string GetPropertyJson()
+    {
+        return _propertyJson;
+    }
+
+    internal void AddPropertyToCollection(string propertyKey, object propertyValue)
+    {
+        _propertiesList.Add(propertyKey, propertyValue);
+    }
+}
+```
+</details>
+
+
+<details>
+<summary>WritablePropertyResponse</summary>
+
+```csharp
+public class WritablePropertyResponse
+{
+    private readonly PropertyConvention _propertyConvention;
+
+    /// <summary>
+    /// Convenience constructor for specifying only the property value.
+    /// </summary>
+    /// <param name="propertyValue">The unserialized property value.</param>
+    /// <param name="propertyConvention"></param>
+    public WritablePropertyResponse(object propertyValue, PropertyConvention propertyConvention)
```

Comment:
the writable property response contains user defined payload "Value", so we allow the serialization to be customizable.